File: sem2/cpp/dll.cpp
#include <bits/stdc++.h>

using namespace std;
class node {

public: 
	int data;
	node* prev;
	node* next;
};

node* makehead(int data) {
	node* root = new node();
	root->data = data;
	root->prev = NULL;
	root->next = NULL;
	return root;
}

void printlist(node* head) {
	node* tmp = head;
	while(tmp != NULL) {
		cout << tmp->data << " ";
		tmp = tmp->next;
	}
}

void addnode(node* head, int data) {
	node* tmp = head;
	while(tmp->next != NULL)
		tmp = tmp->next;
	node* cur = new node();
	cur->data = data;
	cur->prev = tmp;
	tmp->next = cur;
	cur->next = NULL;
}

void addbefore(node* &head, int data) {
	node* cur = new node();
	cur->data = data;
	head->prev = cur;
	cur->next = head;
	cur->prev = NULL;
	head = cur;
}

void printrev(node* head) {
	node* tmp = head;
	while(tmp->next != NULL)
		tmp = tmp->next;
	while(tmp!=NULL) {
		cout << tmp->data << " ";
		tmp = tmp->prev;
	}
}

void deletenode(node* &head, int index) {
	int cnt = 1; node* tmp = head;
	if(index == 1) {
		tmp = head->next;
		tmp->prev = NULL;
		head->next = NULL;
		head = tmp;
	}
	else {
		while(cnt != index-1) {
			tmp = tmp->next;
			cnt++;
		}
		node* cur = tmp->next;
		if(cur->next == NULL) {
			tmp->next = NULL;
			cur->prev = NULL;
			delete cur;
			return;
		}
		tmp->next = cur->next;
		cur->next->prev = cur->prev;
		cur->next = NULL; cur->prev = NULL;
		delete cur;
	}
}

int main() {
	node* root = makehead(5);
	addnode(root, 4); addnode(root, 3); addnode(root, 2);
	addbefore(root, 1);
	printrev(root); cout << '\n';
	deletenode(root,5);
	printlist(root);
	return 0;
}

File: sem2/python/dll.py
#!/usr/bin/python3

class node:
    def __init__(self, data):
        self.data = data
        self.next = None
        self.prev = None

class dll:
    def __init__(self):
        self.head = None;
    def addhead(self,data):
        self.head = node(data)
    def addnode(self, data):
        tmp = self.head
        while(tmp.next != None):
            tmp = tmp.next
        tmp.next = node(data)
        prev_node = tmp
        tmp = tmp.next
        tmp.prev = prev_node
    def printlist(self):
        tmp = self.head
        while(tmp != None):
            print(tmp.data, end = " ")
            tmp = tmp.next
    def printrev(self):
        tmp = self.head
        while(tmp.next != None):
            tmp = tmp.next
        while(tmp):
            print(tmp.data, end = " ")
            tmp = tmp.prev
    def deletenode(self,index):
        if index == 1:
            tmp = self.head
            self.head = tmp.next
            self.head.prev = None
            tmp.next = None
        else:
            cnt = 1
            tmp = self.head
            while(tmp != None and cnt < index-1):
                tmp = tmp.next
                cnt += 1
            cur = tmp.next
            cur.next.prev = cur.prev
            tmp.next = cur.next
            cur.prev = None
            cur.next = None

d = dll()
d.addhead(2)
d.addnode(3)
d.addnode(4)
d.addnode(5)
d.addnode(6)
d.deletenode(3)
d.printlist()
d.printrev()

